File: starlore-py/app/services/knowledge_chunk_service.py
"""与 Java 后端一致的结构化知识切片服务。"""

import math
import re

from sqlalchemy import delete, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.article import Article
from app.models.article_chunk import ArticleChunk

TARGET_TOKENS = 600
MAX_TOKENS = 800
OVERLAP_TOKENS = 80

_HEADING = re.compile(r"^#{1,6}\s+.+")
_SENTENCE_BOUNDARY = re.compile(r"(?<=[。！？；.!?;])|(?=\n)|(?<=\n)")


def estimate_tokens(text: str | None) -> int:
    """与 Java 端保持同一估算口径：CJK 约 1 token，其余约 4 字符 1 token。"""
    if not text or not text.strip():
        return 0
    cjk = sum(
        1
        for char in text
        if "\u3400" <= char <= "\u9fff"
        or "\u3040" <= char <= "\u30ff"
        or "\uac00" <= char <= "\ud7af"
    )
    return cjk + max(1, math.ceil((len(text) - cjk) / 4))


def split_text(text: str | None) -> list[str]:
    normalized = (text or "").replace("\r\n", "\n").replace("\r", "\n").strip()
    if not normalized:
        return []

    chunks: list[str] = []
    current = ""
    for unit in _structural_units(normalized):
        for part in _split_oversized(unit):
            candidate = f"{current}\n\n{part}".strip() if current else part.strip()
            if current and estimate_tokens(candidate) > MAX_TOKENS:
                completed = current.strip()
                chunks.append(completed)
                current = _overlap_tail(completed)
            current = f"{current}\n\n{part}".strip() if current else part.strip()
            if estimate_tokens(current) >= TARGET_TOKENS:
                completed = current.strip()
                chunks.append(completed)
                current = _overlap_tail(completed)

    remainder = current.strip()
    overlap_only = (
        bool(chunks)
        and estimate_tokens(remainder) <= OVERLAP_TOKENS
        and chunks[-1].endswith(remainder)
    )
    if remainder and not overlap_only and (not chunks or remainder != chunks[-1]):
        chunks.append(remainder)
    return chunks


async def get_chunks(db: AsyncSession, article_id: int) -> list[ArticleChunk]:
    result = await db.execute(
        select(ArticleChunk)
        .where(ArticleChunk.articleId == article_id)
        .order_by(ArticleChunk.chunkIndex)
    )
    return list(result.scalars().all())


async def ensure_chunks(db: AsyncSession, article: Article) -> list[ArticleChunk]:
    chunks = await get_chunks(db, article.id)
    return chunks or await rebuild_chunks(db, article)


async def rebuild_chunks(db: AsyncSession, article: Article) -> list[ArticleChunk]:
    await db.execute(delete(ArticleChunk).where(ArticleChunk.articleId == article.id))
    chunks: list[ArticleChunk] = []
    for index, content in enumerate(split_text(article.content)):
        chunk = ArticleChunk(
            articleId=article.id,
            chunkIndex=index,
            content=content,
            tokenCount=estimate_tokens(content),
            isEnabled=1,
        )
        db.add(chunk)
        chunks.append(chunk)
    await db.flush()
    return chunks


async def delete_chunks(db: AsyncSession, article_id: int) -> None:
    await db.execute(delete(ArticleChunk).where(ArticleChunk.articleId == article_id))


def _structural_units(text: str) -> list[str]:
    units: list[str] = []
    block: list[str] = []
    in_code = False

    def flush() -> None:
        value = "\n".join(block).strip()
        if value:
            units.append(value)
        block.clear()

    for line in text.split("\n"):
        stripped = line.strip()
        if stripped.startswith("```"):
            if not in_code:
                flush()
            in_code = not in_code
            block.append(line)
            if not in_code:
                flush()
        elif not in_code and (not stripped or _HEADING.match(stripped)):
            flush()
            if stripped:
                units.append(stripped)
        else:
            block.append(line)
    flush()
    return units


def _split_oversized(unit: str) -> list[str]:
    if estimate_tokens(unit) <= MAX_TOKENS:
        return [unit]
    parts: list[str] = []
    current = ""
    for sentence in _SENTENCE_BOUNDARY.split(unit):
        if not sentence.strip():
            continue
        if current and estimate_tokens(current + sentence) > MAX_TOKENS:
            parts.append(current.strip())
            current = _overlap_tail(current)
        current += sentence
    if current.strip():
        parts.append(current.strip())

    bounded: list[str] = []
    for part in parts:
        bounded.extend([part] if estimate_tokens(part) <= MAX_TOKENS else _hard_split(part))
    return bounded


def _hard_split(text: str) -> list[str]:
    parts: list[str] = []
    current = ""
    for char in text:
        current += char
        if estimate_tokens(current) >= MAX_TOKENS:
            completed = current.strip()
            parts.append(completed)
            current = _overlap_tail(completed)
    if current.strip():
        parts.append(current.strip())
    return parts


def _overlap_tail(text: str) -> str:
    start = len(text)
    while start > 0 and estimate_tokens(text[start:]) < OVERLAP_TOKENS:
        start -= 1
    tail = text[start:].strip()
    boundaries = [tail.find(mark) for mark in ("。", "！", "？", "\n")]
    boundary = max(boundaries)
    return tail[boundary + 1 :].strip() if 0 <= boundary + 1 < len(tail) else tail

File: starlore-back/src/main/java/com/robin/blogback/service/KnowledgeChunkService.java
package com.robin.blogback.service;

import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.robin.blogback.entity.Article;
import com.robin.blogback.entity.ArticleChunk;
import com.robin.blogback.mapper.ArticleChunkMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.regex.Pattern;

@Service
@RequiredArgsConstructor
public class KnowledgeChunkService {

    private static final int TARGET_TOKENS = 600;
    private static final int MAX_TOKENS = 800;
    private static final int OVERLAP_TOKENS = 80;
    private static final Pattern HEADING = Pattern.compile("^#{1,6}\\s+.+");
    private static final Pattern SENTENCE_BOUNDARY =
            Pattern.compile("(?<=[。！？；.!?;])|(?=\\n)|(?<=\\n)");

    private final ArticleChunkMapper articleChunkMapper;

    public List<ArticleChunk> getChunks(Integer articleId) {
        return articleChunkMapper.selectList(
                new LambdaQueryWrapper<ArticleChunk>()
                        .eq(ArticleChunk::getArticleId, articleId)
                        .orderByAsc(ArticleChunk::getChunkIndex));
    }

    public List<ArticleChunk> ensureChunks(Article article) {
        if (article == null) return Collections.emptyList();
        List<ArticleChunk> chunks = getChunks(article.getId());
        return chunks.isEmpty() ? rebuildChunks(article) : chunks;
    }

    @Transactional
    public List<ArticleChunk> rebuildChunks(Article article) {
        if (article == null || article.getId() == null) return Collections.emptyList();
        articleChunkMapper.delete(
                new LambdaQueryWrapper<ArticleChunk>().eq(ArticleChunk::getArticleId, article.getId()));

        String content = normalize(article.getContent());
        if (content.isBlank()) return Collections.emptyList();

        List<String> pieces = splitMarkdown(content);
        List<ArticleChunk> chunks = new ArrayList<>();
        for (int index = 0; index < pieces.size(); index++) {
            String piece = pieces.get(index).trim();
            if (piece.isEmpty()) continue;
            ArticleChunk chunk = new ArticleChunk();
            chunk.setArticleId(article.getId().longValue());
            chunk.setChunkIndex(index);
            chunk.setContent(piece);
            chunk.setTokenCount(estimateTokens(piece));
            chunk.setIsEnabled(1);
            articleChunkMapper.insert(chunk);
            chunks.add(chunk);
        }
        return chunks;
    }

    public int estimateTokens(String text) {
        if (text == null || text.isBlank()) return 0;
        int cjk = 0;
        int ascii = 0;
        for (int offset = 0; offset < text.length();) {
            int codePoint = text.codePointAt(offset);
            offset += Character.charCount(codePoint);
            if (isCjk(codePoint)) cjk++;
            else ascii++;
        }
        return cjk + Math.max(1, (int) Math.ceil(ascii / 4.0));
    }

    @Transactional
    public void deleteChunks(Integer articleId) {
        if (articleId == null) return;
        articleChunkMapper.delete(
                new LambdaQueryWrapper<ArticleChunk>().eq(ArticleChunk::getArticleId, articleId));
    }

    private List<String> splitMarkdown(String text) {
        List<String> units = structuralUnits(text);
        List<String> chunks = new ArrayList<>();
        StringBuilder current = new StringBuilder();

        for (String unit : units) {
            for (String part : splitOversized(unit)) {
                if (!current.isEmpty()
                        && estimateTokens(current + "\n\n" + part) > MAX_TOKENS) {
                    String completed = current.toString().trim();
                    chunks.add(completed);
                    current = new StringBuilder(overlapTail(completed));
                }
                if (!current.isEmpty()) current.append("\n\n");
                current.append(part.trim());
                if (estimateTokens(current.toString()) >= TARGET_TOKENS) {
                    String completed = current.toString().trim();
                    chunks.add(completed);
                    current = new StringBuilder(overlapTail(completed));
                }
            }
        }
        String remainder = current.toString().trim();
        boolean overlapOnly = !chunks.isEmpty()
                && estimateTokens(remainder) <= OVERLAP_TOKENS
                && chunks.get(chunks.size() - 1).endsWith(remainder);
        if (!remainder.isEmpty() && !overlapOnly
                && (chunks.isEmpty() || !remainder.equals(chunks.get(chunks.size() - 1)))) {
            chunks.add(remainder);
        }
        return chunks;
    }

    private List<String> structuralUnits(String text) {
        List<String> units = new ArrayList<>();
        StringBuilder block = new StringBuilder();
        boolean inCode = false;
        for (String line : text.split("\\n", -1)) {
            String trimmed = line.trim();
            if (trimmed.startsWith("```")) {
                if (!inCode) flush(block, units);
                inCode = !inCode;
                block.append(line).append('\n');
                if (!inCode) flush(block, units);
            } else if (!inCode && (trimmed.isEmpty() || HEADING.matcher(trimmed).matches())) {
                flush(block, units);
                if (!trimmed.isEmpty()) units.add(trimmed);
            } else {
                block.append(line).append('\n');
            }
        }
        flush(block, units);
        return units;
    }

    private List<String> splitOversized(String unit) {
        if (estimateTokens(unit) <= MAX_TOKENS) return List.of(unit);
        List<String> result = new ArrayList<>();
        StringBuilder part = new StringBuilder();
        for (String sentence : SENTENCE_BOUNDARY.split(unit)) {
            if (sentence.isBlank()) continue;
            if (!part.isEmpty() && estimateTokens(part + sentence) > MAX_TOKENS) {
                result.add(part.toString().trim());
                part = new StringBuilder(overlapTail(part.toString()));
            }
            part.append(sentence);
        }
        if (!part.isEmpty()) result.add(part.toString().trim());

        List<String> bounded = new ArrayList<>();
        for (String value : result) {
            if (estimateTokens(value) <= MAX_TOKENS) bounded.add(value);
            else bounded.addAll(hardSplit(value));
        }
        return bounded;
    }

    private List<String> hardSplit(String text) {
        List<String> result = new ArrayList<>();
        StringBuilder part = new StringBuilder();
        for (int offset = 0; offset < text.length();) {
            int codePoint = text.codePointAt(offset);
            offset += Character.charCount(codePoint);
            part.appendCodePoint(codePoint);
            if (estimateTokens(part.toString()) >= MAX_TOKENS) {
                String completed = part.toString().trim();
                result.add(completed);
                part = new StringBuilder(overlapTail(completed));
            }
        }
        String remainder = part.toString().trim();
        if (!remainder.isEmpty()) result.add(remainder);
        return result;
    }

    private String overlapTail(String text) {
        if (text.isBlank()) return "";
        int start = text.length();
        while (start > 0 && estimateTokens(text.substring(start)) < OVERLAP_TOKENS) {
            start = text.offsetByCodePoints(start, -1);
        }
        String tail = text.substring(start).trim();
        int boundary = Math.max(
                Math.max(tail.indexOf('。'), tail.indexOf('！')),
                Math.max(tail.indexOf('？'), tail.indexOf('\n')));
        return boundary >= 0 && boundary + 1 < tail.length()
                ? tail.substring(boundary + 1).trim()
                : tail;
    }

    private void flush(StringBuilder block, List<String> units) {
        String value = block.toString().trim();
        if (!value.isEmpty()) units.add(value);
        block.setLength(0);
    }

    private String normalize(String text) {
        return text == null ? "" : text.replace("\r\n", "\n").replace('\r', '\n').trim();
    }

    private boolean isCjk(int codePoint) {
        Character.UnicodeScript script = Character.UnicodeScript.of(codePoint);
        return script == Character.UnicodeScript.HAN
                || script == Character.UnicodeScript.HIRAGANA
                || script == Character.UnicodeScript.KATAKANA
                || script == Character.UnicodeScript.HANGUL;
    }
}
